server/jobs_controller.go

```diff
@@ -0,0 +1,201 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/logging"
+)
+
+// Concurrency model in this controller:
+//
+// - We keep a buffer of all the output so far for each PR
+// - We keep a set of clients for each PR
+// - When a new client is added, the entire buffer is sent through so
+//   it is "up to date"
+// - When a new line of output is added, we send it to all current
+//   clients and then add it to the buffer
+// - Changes to the buffer or set of clients are wrapped in a mutex
+//
+// This maintains the invariant that whenever a function call returns,
+// all of the clients have seen every message that is in the buffer.
+//
+// When we get a message that tells us to clear the buffer, we just do
+// it, without telling any clients about it. This won't have any
+// effect on the frontend, except that newly connected clients won't
+// get the old messages sent to them, because the server has
+// "forgotten" about them. We clear the buffer for a PR when starting
+// and finishing processing that PR, because after the processing has
+// finished the output is posted to GitHub anyway and it doesn't need
+// to be persisted in the Atlantis web UI.
+
+type JobsController struct {
+	AtlantisVersion     string
+	AtlantisURL         *url.URL
+	Logger              *logging.SimpleLogger
+	JobTemplate         TemplateWriter
+	TerraformOutputChan <-chan *models.TerraformOutputLine
+
+	logBuffers map[string][]string
```

Comment:
can see this easily being customized to be a pluggable "backend" allowing us to actually persist these.